src/home/homework-luks.c

```diff
@@ -1930,23 +1974,69 @@ static int make_partition_table(
         if (r < 0)
                 return log_error_errno(r, "Failed to parse disk label UUID: %m");
 
-        r = fdisk_get_partition(c, 0, &q);
+        r = query_partition_offset_and_size(c, 0, ret_offset, ret_size);
         if (r < 0)
-                return log_error_errno(r, "Failed to read created partition metadata: %m");
+                return r;
 
-        assert(fdisk_partition_has_start(q));
-        offset = fdisk_partition_get_start(q);
-        if (offset > UINT64_MAX / 512U)
-                return log_error_errno(SYNTHETIC_ERRNO(ERANGE), "Partition offset too large.");
+        *ret_disk_uuid = disk_uuid;
 
-        assert(fdisk_partition_has_size(q));
-        size = fdisk_partition_get_size(q);
-        if (size > UINT64_MAX / 512U)
-                return log_error_errno(SYNTHETIC_ERRNO(ERANGE), "Partition size too large.");
+        return 0;
+}
 
-        *ret_offset = offset * 512U;
-        *ret_size = size * 512U;
-        *ret_disk_uuid = disk_uuid;
+static int update_partition_table(
+                int fd,
+                const char *label,
+                int partno,
+                sd_id128_t partition_uuid,
+                uint64_t *ret_offset,
+                uint64_t *ret_size,
+                sd_id128_t *ret_disk_uuid) {
+        uint64_t partition_offset = 0, partition_size = 0;
+        _cleanup_(fdisk_unref_contextp) struct fdisk_context *c = NULL;
+        _cleanup_(fdisk_unref_partitionp) struct fdisk_partition *p = NULL;
+        _cleanup_free_ char *path = NULL;
+        int r;
+
+        c = fdisk_new_context();
+        if (!c)
+                return log_oom();
+
+        if (asprintf(&path, "/proc/self/fd/%i", fd) < 0)
+                return log_oom();
+
+        r = fdisk_assign_device(c, path, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open device: %m");
```

Comment:
these three operations here are the same as for make_partition_table(), and as I see now we have the same code elsewhere too, so maybe let's turn this into a proper helper:

→ #25503

can you rebase on top, and replace these three lines?

Reply:
Thanks ! I rebased (on main now that it's been merged) and now use this helper